### PR TITLE
Removed NodeJS event emitter in common library.

### PR DESCRIPTION
**Pull request by lukehb**
## Relevant components:
- [x] Signalling server
- [x] Common library
- [x] Frontend library
- [x] Frontend UI library

## Problem statement:
The `Common` library contained a NodeJS type, `EventEmitter` this was causing it to fail when used in the browser. This was not a problem previously when we bundled at the library build step, but since https://github.com/EpicGamesExt/PixelStreamingInfrastructure/pull/324 have been a few dependencies that do not work well in the browser without bundling.

This issue was reported in: https://github.com/EpicGamesExt/PixelStreamingInfrastructure/issues/328 and impacts the latest version of UE 5.5 frontend library.

## Solution
This PR solves the issue by making a browser compatible EventEmitter type using EventTarget under the hood.

## Documentation
N/A

## Test Plan and Compatibility
I have tested by:
- Running the unit tests
- Running the frontend tests
- Running with an actual UE instance
- Running the signalling server

### CODE REVIEW

File: Common/src/Event/EventEmitter.ts [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+class PixelStreamingEventListener implements EventListenerObject {
+    private _callback: Function;
+    private _args: any[] = [];
+
+    constructor(callback: Function) {
+        this._callback = callback;
+    }
+
+    handleEvent(_evt: Event): void {
```
Comment (mcottontensor) [inline Common/src/Event/EventEmitter.ts:17]: Should the event not be passed to the callback?

File: Common/src/Event/EventEmitter.ts [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        this._callback(...this._args);
+    }
+
+    set args(value: any) {
+        this._args = value;
```
Comment (mcottontensor) [inline Common/src/Event/EventEmitter.ts]: I'm guessing when you assign an any[] parameter to this.args it comes through as any[] which is also covered by any.. Should this be more explicit?

File: Extras/FrontendTests/tests/mouse.spec.ts [hunk 3]
```diff
@@ -108,7 +114,7 @@ test('Test locked mouse movement', {
         console.log("Player: ", ...args);
     });
 
-    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true&HoveringMouse=false`);
+    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=false&HoveringMouse=false`);
```
Comment (lukehb) [inline Extras/FrontendTests/tests/mouse.spec.ts]:
```suggestion
    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true&HoveringMouse=false`);
```

File: Extras/FrontendTests/tests/mouse.spec.ts [hunk 4]
```diff
@@ -238,7 +244,7 @@ test('Test mouse input after resizing. Hover mouse.', {
     //     console.log("Player: ", ...args);
     // });
 
-    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true&HoveringMouse=true`);
+    await page.goto(`/?StreamerId=${streamerId}&HoveringMouse=true`);
```
Comment (lukehb) [inline Extras/FrontendTests/tests/mouse.spec.ts]:
```suggestion
    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true&HoveringMouse=true`);
```

File: Extras/FrontendTests/tests/mouse.spec.ts [hunk 5]
```diff
@@ -296,7 +308,7 @@ test('Test mouse input after resizing. locked mouse.', {
     //     console.log("Player: ", ...args);
     // });
 
-    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true&HoveringMouse=false`);
+    await page.goto(`/?StreamerId=${streamerId}&HoveringMouse=false`);
```
Comment (lukehb) [inline Extras/FrontendTests/tests/mouse.spec.ts]:
```suggestion
    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true&HoveringMouse=false`);
```

File: Extras/JSStreamer/package.json [hunk 6]
```diff
@@ -7,7 +7,7 @@
   "module": "build/pixelstreamingjsstreamer.js",
   "scripts": {
     "develop": "webpack-dev-server --mode development",
-    "build": "tsc --project tsconfig.json && webpack --mode production",
+    "build": "tsc --project tsconfig.json && webpack --mode development",
```
Comment (lukehb) [inline Extras/JSStreamer/package.json:10]: Didn't really make sense to minify/obscure this code as it is only used in dev.

File: Extras/JSStreamer/webpack.config.js [hunk 7]
```diff
@@ -45,7 +45,7 @@ module.exports = {
             },
         },
         static: path.join(__dirname, 'dist'),
-        compress: true,
+        compress: false,
```
Comment (lukehb) [inline Extras/JSStreamer/webpack.config.js:48]: Didn't make sense to compress this code as we only use it for testing/dev.

File: Frontend/library/src/Util/EventEmitter.ts [hunk 8]
```diff
@@ -581,7 +581,7 @@ export type PixelStreamingEvent =
     | PlayerCountEvent
     | WebRtcTCPRelayDetectedEvent;
 
-export class EventEmitter extends EventTarget {
+export class PixelStreamingEventEmitter extends EventTarget {
```
Comment (lukehb) [inline Frontend/library/src/Util/EventEmitter.ts:584]: We already had a class called EventEmitter, so I renamed it here. This seemed like the lesser impactful rename as opposed to renaming the one in common.

File: Extras/FrontendTests/package.json [hunk 9]
```diff
@@ -4,13 +4,14 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "test": "npx playwright test"
+    "test": "npx playwright test",
+    "build": ""
```
Comment (lukehb) [inline Extras/FrontendTests/package.json:8]: You may notice a few empty build scripts in this commit.

Those were added to enable running `npm run build` in the root of the repo and building all workspaces - handy.

File: SignallingWebServer/Dockerfile [hunk 10]
```diff
@@ -1,6 +1,7 @@
 # Use the current Long Term Support (LTS) version of Node.js
 FROM node:lts
 # Copy the signalling server and frontend source code from the build context
+COPY package.json package.json
```
Comment (lukehb) [inline SignallingWebServer/Dockerfile:4]: This was needed in the Dockerfile to make it use the new NPM workspaces way of doing things.

File: package.json [hunk 11]
```diff
@@ -15,5 +15,8 @@
 		"SignallingWebServer",
 		"SS_Test"
 	],
-	"private": true
+	"private": true,
+	"scripts": {
+		"build": "npm run build --ws"
+	}
```
Comment (lukehb) [inline package.json:21]: This lets us run `npm run build` in the repo root and build everything.

File: Common/src/Event/EventEmitter.ts (hunk 1 above)
Comment (mcottontensor) [reply Common/src/Event/EventEmitter.ts:17]: Still questioning this one.

**Review by mcottontensor — APPROVED** [review]

File: Common/src/Event/EventEmitter.ts (hunk 1 above)
Comment (lukehb) [reply Common/src/Event/EventEmitter.ts:17]: passing that Event to the downstream listener would actually introduce different behaviour than what node js was doing with its event emitter

It has no meaningful information attached to it and is really an internal implementation detail, exposing it to the listener would be a leaky abstraction considering the point was the make something that looks and smells likes NodeJS event emitter but actually uses EventTarget under the hood.

File: Common/src/Event/EventEmitter.ts (hunk 2 above)
Comment (lukehb) [reply Common/src/Event/EventEmitter.ts]: Has been updated now.

**Pull request merged**